source/common/upstream/edf_scheduler.h

```diff
@@ -64,7 +66,92 @@ template <class C> class EdfScheduler : public Scheduler<C> {
+      return calculate_weight(entry) + 0.00000000001;
+    };
+
+    // Let weights {w_1, w_2, ..., w_N} be the per-entry weight where (w_i > 0),
+    // W = sum(w_i), and P be the number of times to "pick" from the scheduler.
+    // Let p'_i = floor(P * w_i/W), then the number of times each entry is being
+    // picked is p_i >= p'_i. Note that 0 <= P - sum(p'_i) < N.
+    //
+    // The following code does P picks, by first emulating p'_i picks for each
+    // entry, and then executing the leftover P - sum(p'_i) picks.
+    double weights_sum = std::accumulate(
+        entries.cbegin(), entries.cend(), 0.0,
+        [&aug_calculate_weight](double sum_so_far, const std::shared_ptr<C>& entry) {
+          return sum_so_far + aug_calculate_weight(*entry);
+        });
+    std::vector<uint32_t> floor_picks;
+    floor_picks.reserve(entries.size());
+    std::transform(entries.cbegin(), entries.cend(), std::back_inserter(floor_picks),
+                   [picks, weights_sum, &aug_calculate_weight](const std::shared_ptr<C>& entry) {
+                     // Getting the lower-bound by casting to an integer.
+                     return static_cast<uint32_t>(aug_calculate_weight(*entry) / weights_sum *
+                                                  picks);
+                   });
+
+    // Pre-compute the priority-queue entries to use an O(N) initialization c'tor.
+    std::vector<EdfEntry> scheduler_entries;
+    scheduler_entries.reserve(entries.size());
+    uint32_t order_offset = 0;
+    uint32_t picks_so_far = 0;
+    double max_pick_time = 0.0;
+    // Emulate a per-entry add to a deadline that is applicable to N picks.
+    for (size_t i = 0; i < entries.size(); ++i) {
+      // Add the entry with c'_i picks. As there were c'_i picks, the entry's
+      // next deadline is (c'_i + 1) / w_i.
+      const double weight = aug_calculate_weight(*entries[i]);
+      const double pick_time = floor_picks[i] / weight;
+      const double deadline = pick_time + 1.0 / weight;
+      EDF_TRACE("Insertion {} in queue with emualted {} picks, deadline {} and weight {}.",
+                static_cast<const void*>(entries[i].get()), floor_picks[i], deadline, weight);
+      scheduler_entries.emplace_back(EdfEntry{deadline, order_offset++, entries[i]});
+      max_pick_time = std::max(max_pick_time, pick_time);
+      picks_so_far += floor_picks[i];
+    }
+    // The scheduler's current_time_ needs to be the largest time that some entry was picked.
+    EdfScheduler<C> scheduler(std::move(scheduler_entries), max_pick_time, order_offset);
+    ASSERT(scheduler.queue_.top().deadline_ >= scheduler.current_time_);
+
+    // Left to do some picks, execute them one after the other.
+    EDF_TRACE("Emulated {} picks in init step, {} picks remaining for one after the other step",
+              picks_so_far, picks - picks_so_far);
+    for (; picks_so_far < picks; ++picks_so_far) {
+      scheduler.pickAndAdd(aug_calculate_weight);
```

Comment:
Should you be passing in `aug_calculate_weight` or `calculate_weight` here? You've already perturbed the weights, so at this point you should use the original weight calculation, right?